### ISSUE TITLE
Long running workflows with time based activities

### ISSUE THREAD

**Issue by mustafashabib** [aws/aws-flow-ruby#106]
Hi again

Just hoping we can get some guidance on our workflow and how to implement it.

A simple example of our workflow can be described as:
1. begin the workflow
2. if the time from the first start of the workflow to the current time is less than **7 days**, continue, else perform a "workflow didn't succeed" task that does some stuff (send an email in this case, not that it matters)
3. if the workflow is executing within the time window allocated for it, check for the status of something on our backend
   - if the status is "true" (for example), process a new task
   - if the status is "false" (for example), start the workflow again 

I think that the periodic example in the sample code basically defines the pattern we should use - but we're still confused as to whether starting a new workflow using `continue_as_new` is the right construct to use here. 

If it matters, this workflow is expected to be kicked off by a parent workflow in the typical case and the parent workflow must continue with further task processing until a point where it must wait for this child workflow to complete. 

Some (bogus?) test code I wrote based on the periodic example can be found here: https://github.com/mustafashabib/aws-flow-ruby-experiments

Thanks for any help.

M

**Comment by mustafashabib** [comment]
To add some context, the output I see when I kick off the workflow is:

If I use [the workflow modeled after the periodic sample](https://github.com/mustafashabib/aws-flow-ruby-experiments/blob/master/referral_status_check_workflow.rb) I get the output:

``` bash
{:activity_name=>:check_referral, :activity_prefix=>"ReferralStatusCheckActivity", :referral_status_poll_time=>5, :wait_for_completion=>false, :continue_as_new_secs=>1, :workflow_seconds_remaining=>60}
Workflow has started
Checking if referral status is not pending
check_referral
60s remaining
check_referral
waiting 5s before trying again
60s remaining
check_referral
waiting 5s before trying again
check_referral
Still Pending
Result was found as . Workflow completed.
```

If I use [the workflow using a for loop instead](https://github.com/mustafashabib/aws-flow-ruby-experiments/blob/master/referral_status_check_workflow_alternate.rb) it seems to work as expected in all the cases I defined (status changed successfully or status remained `PENDING` for the entire time allowed for this workflow.)

I'm not sure why the first one doesn't work and what I'm missing conceptually. 

Thanks for any help.

**Comment by mjsteger** [comment]
So, it sounds like continue_as_new may be what you want to use, but I'd like to clarify how it works and see if it matches your use case. What continue_as_new does is set a flag such that, on workflow completion, the workflow will continue as new instead of simply completing. It does not immediately exit the workflow, and if there are still child processes(activities, workflows) being processed, the parent workflow will not continue_as_new if it is still waiting on them. I'd be a little worried about managing the complexity of a parent workflow waiting on a child workflow that can continue as new personally.

The reason that you get the above result is that the return value of the recursive call `check_status_referral` is nil if you keep getting PENDING and then reach `duration > options[:continue_as_new_secs]`(as the code will fall to the end of the if at line 101, no else case defined, and thus return nil, which will then be recursively returned up). Adding a line like so:

<pre>
  check_referral_status(options, start_time)
  <b>status_result</b>
</pre>

in referral_status_check_workflow.rb:116 will make it continue as new correctly. Running it with that code change:

```
{:activity_name=>:check_referral, :activity_prefix=>"ReferralStatusCheckActivity", :referral_status_poll_time=>5, :wait_for_completion=>false, :continue_as_new_secs=>12, :workflow_seconds_remaining=>60}
Workflow has started
Checking if referral status is not pending
check_referral
60s remaining
check_referral
waiting 5s before trying again
60s remaining
check_referral
waiting 5s before trying again
check_referral
Still Pending
23.58400011062622s left
Workflow is continuing as new
{:activity_name=>:check_referral, :activity_prefix=>"ReferralStatusCheckActivity", :referral_status_poll_time=>5, :wait_for_completion=>false, :continue_as_new_secs=>12, :workflow_seconds_remaining=>23.58400011062622}
Workflow has started
Checking if referral status is not pending
check_referral
23.58400011062622s remaining
check_referral
waiting 5s before trying again
23.58400011062622s remaining
check_referral
waiting 5s before trying again
check_referral
23.58400011062622s remaining
check_referral
waiting 5s before trying again
check_referral
waiting 5s before trying again
23.58400011062622s remaining
check_referral
waiting 5s before trying again
check_referral
waiting 5s before trying again
check_referral
Still Pending
10.960000038146973s left
Workflow is continuing as new
{:activity_name=>:check_referral, :activity_prefix=>"ReferralStatusCheckActivity", :referral_status_poll_time=>5, :wait_for_completion=>false, :continue_as_new_secs=>12, :workflow_seconds_remaining=>10.960000038146973}
Workflow has started
Checking if referral status is not pending
check_referral
10.960000038146973s remaining
check_referral
waiting 5s before trying again
10.960000038146973s remaining
check_referral
waiting 5s before trying again
check_referral
10.960000038146973s remaining
check_referral
waiting 5s before trying again
check_referral
waiting 5s before trying again
10.960000038146973s remaining
check_referral
waiting 5s before trying again
check_referral
waiting 5s before trying again
check_referral
Still Pending
-10.360999822616577s left
No time left and result was still pending, time to send an email.
```

**Comment by mustafashabib** [comment]
Wow thanks so much for your help @mjsteger and explanation around `continue_as_new`!
